src/Illuminate/Support/Facades/Date.php

```diff
@@ -0,0 +1,178 @@
+ * @method static string singularUnit(string $unit)
+ * @method static \Illuminate\Support\Carbon today($tz = null)
+ * @method static \Illuminate\Support\Carbon tomorrow($tz = null)
+ * @method static void useMonthsOverflow($monthsOverflow = true)
+ * @method static \Illuminate\Support\Carbon useStrictMode($strictModeEnabled = true)
+ * @method static void useYearsOverflow($yearsOverflow = true)
+ * @method static \Illuminate\Support\Carbon yesterday($tz = null)
+ */
+class Date extends Facade
+{
+    /**
+     * Date instance class name.
+     *
+     * @var string
+     */
+    protected static $className = Carbon::class;
+
+    /**
+     * Date interceptor.
+     *
+     * @var callable
+     */
+    protected static $interceptor;
+
+    /**
+     * Get the registered name of the component.
+     *
+     * @return string
+     *
+     * @throws \RuntimeException
+     */
+    protected static function getFacadeAccessor()
+    {
+        return 'date';
+    }
+
+    /**
+     * Change the class to use for date instances.
+     *
+     * @param string $className
+     */
+    public static function use(string $className)
+    {
+        if (! (class_exists($className) && (in_array(DateTimeInterface::class, class_implements($className)) || method_exists($className, 'instance')))) {
+            throw new InvalidArgumentException(
+                'Date class must implement a public static instance(DateTimeInterface $date) method or implements DateTimeInterface.'
+            );
+        }
+
+        static::$className = $className;
+    }
+
+    /**
+     * Set an interceptor for each date (Carbon instance).
+     *
+     * @param callable $interceptor
+     */
+    public static function intercept(callable $interceptor)
+    {
+        static::$interceptor = $interceptor;
+    }
+
+    /**
+     * Handle dynamic, static calls to the object.
+     *
+     * @param  string  $method
+     * @param  array   $args
+     *
+     * @return mixed
+     *
+     * @throws \RuntimeException
+     */
+    public static function __callStatic($method, $args)
+    {
+        try {
+            if (static::getFacadeRoot()) {
+                return parent::__callStatic($method, $args);
+            }
+        } catch (ReflectionException $exception) {
+            // continue
+        }
+
+        if (method_exists(static::$className, $method)) {
+            $date = static::$className::$method(...$args);
+        } else {
+            /** @var Carbon $date */
+            $date = Carbon::$method(...$args);
+            if (method_exists(static::$className, 'instance')) {
+                $date = static::$className::instance($date);
+            } else {
+                $date = new static::$className($date->format('Y-m-d H:i:s.u'), $date->getTimezone());
+            }
+        }
+
+        if (static::$interceptor) {
+            return call_user_func(static::$interceptor, $date);
+        }
+
+        return $date;
+    }
```

Comment:
What don't you understand here?

It's about giving different possibilities to the user:
- if Facade swap used, try to use it
- else:
  - try to call the method if it exists on the class (Carbon by default, but it can be changed with `use`)
  - else call it on Carbon original class then create the new object with it
    - we first try to create the new object using `instance`
    - if `instance` method does not exist, create it using constructor assuming the constructor is compatible with DateTime (matches the `if instance method or implements DateTimeInterface` in `use` method)
  - in both case apply the interceptor if specified by the user

There's nothing more.

Reply:
What's the reason to have `static::$className` feature  *and* the fallback to `Carbon::` in `__callStatic`?

Why don't we go only for `static::$className`, why the fallback mechanism?

Reply:
Suppose you have `Date::use(SomeClass::class)`, then in the framework, Laravel calls `Date::someMethod()` if `someMethod` exists in `Carbon` but not in `SomeClass`, then we will do `Carbon::someMethod`, then create a `SomeClass` object from it. This way, inside Laravel we can continue to use `Date` as if it is the `Carbon` class no matter what other class would choose the end-user.

Reply:
Not agree. In my opinion, Laravel devs should benefit of all features of Carbon to create dates, and end-users should just have a way to convert those instances into different objects. But Laravel should not care about what class the user chose. When you have `Date::createFromTimestamp()` in the framework code (not the application code), it should not requires `createFromTimestamp` to be implemented in the class chosen by the end-user.

Reply:
That's exactly where we disagree. If I configure the facade to not use Carbon, it should never use Carbon under no circumstances whatsoever.

Reply:
I'm not sure you get the feature, it's not a facade we will expose to the end-user. The user will not do `\Illuminate\Supports\Facades\Date::something()`, there is no need for the user to change his habits, if he used `now()` function or `\Illuminte\Support\Carbon`, he will just keep using them as before. If he want to use an other class, then he will do as before too `CarbonImmutable::now()` then if he run `Date::use(CarbonImmutable::class)`, he will get `CarbonImmutable` instances when calling `now()` too and when retrieving a date from the DB. That's the purpose. The only methods the end-user will need use from this facade is `use` or `intercept`.

Then, I'm not going to prevent Laravel developers to use Carbon internally, that would be dummy. And if some end-user want raw `DateTime` objects, of course it does not mean Laravel will use `DateTime` internally.

The `Date` facade is the common endpoint between the user desired class (could be any DateTimeInterface compatible class) and the dates created internally that can be used by the user and internally, dates will still be created with Carbon, no matter what the end-user want to work with.

Reply:
Just try it in a real app, run `Date::use(DateTime::class)` then get a date such as `$user->created_at`, you will understand why `$date = static::$className::$method(...$args);` is not enough. But if you find a easier solution that does not break **internal** usage of Carbon (and future usages) and still allow to custom the date type for the end-user, I'm open.